File: AlbumView.rb
#
#  album_view.rb
#  ManiaDB
#
#  Created by Appledelhi on 11/16/08.
#  Copyright (c) 2008 Appledelhi. All rights reserved.
#

include OSX

class AlbumView < NSView
  attr_reader :tracks
  attr_writer :imageURL
  
  def initWithFrameActive frame, active
    initWithFrame frame
    setAutoresizingMask NSViewWidthSizable

    @delegate = NSApp.delegate

    button_size = frame[3] - 10
    if active
      @image = NSButton.alloc.initWithFrame [5, 10, button_size, button_size]
      @image.setBezelStyle NSRegularSquareBezelStyle
      @image.setTitle ""
      @image.setTarget self
      @image.setAction "imageClicked:"
    else
      @image = NSImageView.alloc.initWithFrame [5, 10, button_size, button_size]
      @image.setImageFrameStyle NSImageFrameGroove
    end
    @image.cell.setImageScaling NSImageScaleProportionallyUpOrDown
    addSubview @image
    
    @imageSize = addLabel 5, -10, button_size, NSMiniControlSize, nil
    @imageSize.setAlignment NSCenterTextAlignment
    
    x = 5 + button_size + 5
    y = 5 + button_size - 11
    w = frame[2] - x - 5 - 45
    @album = addLabel x, y, w, NSRegularControlSize, "albumClicked:"
    @all = addButton("All", x + w + 5, y + 2, 40, NSMiniControlSize, "allClicked:") if active
    
    y -= 22
    @artist = addLabel x, y, 295, NSSmallControlSize, "artistClicked:"
    @year = addLabel x + 300, y, 50, NSSmallControlSize, "yearClicked:"
    
    @tracks = []
    y -= 16
    for i in 0..3 do
      makeColumn x + (180 + 5) * i, y, 9 * i
    end
    @index = 0
    @score = 0
    
    return self
  end
  
  def image= image
    @image.setImage image
    if image and image.size.width > 0
      @imageSize.setStringValue "#{image.size.width.to_i}x#{image.size.height.to_i}"
    else
      @imageSize.setStringValue ""
    end
    @image.setNeedsDisplay(true)
  end
  
  def album= title
    @album.setStringValue title
    @album.sizeToFit
    frame = @album.frame
    @all.setFrameOrigin([frame.origin.x + frame.size.width + 10, frame.origin.y - 3]) if @all
  end
  
  def artist= artist
    @artist.setStringValue artist
  end
  
  def year= year
    @year.setStringValue year == 0 ? "" : year.to_s
  end
  
  def addTrack disc, track, title, score
    str = disc ? "#{disc}-#{track}. #{title}" : "#{track}. #{title}"
    if @tracks[@index]
      @tracks[@index].setStringValue str
      @tracks[@index].setToolTip(str)
      if score > @score
        @tracks[@score_index].setTextColor NSColor.whiteColor if @score_index
        @tracks[@index].setTextColor NSColor.greenColor
        # @tracks[@index].setStringValue("#{str} (#{score})")
        @track_number = track.to_i
        @song = title
        @score = score
        @score_index = @index
      end
      @index += 1
    end
  end

  def trackNumber= t
    @track_number = t
    if @track_number > 0
      if @song
        @tracks[0].setStringValue("#{@track_number}. #{@song}")
      else
        @tracks[0].setStringValue("#{@track_number}. ")
      end
    elsif @song
      @tracks[0].setStringValue(@song)
    end
  end
  
  def song= song
    @song = song
    @tracks[0].setStringValue @track_number > 0 ? "#{@track_number}. #{@song}" : @song
  end
  
  def addLabel x, y, w, size, action, action2=nil
    label = MyTextField.alloc.initWithFrame [x, y, w, 20]
    label.setAutoresizingMask NSViewWidthSizable
    label.setStringValue ""
    label.setEditable false
    label.setBezeled false
    label.setDrawsBackground false
    label.setTextColor(NSColor.whiteColor)
    label.setTarget self
    label.setAction action
    label.action2 = action2
    fontSize = NSFont.systemFontSizeForControlSize size
    cell = label.cell
    font = NSFont.fontWithName_size(cell.font.fontName, fontSize)
    cell.setFont font
    cell.setControlSize size
    cell.setLineBreakMode NSLineBreakByTruncatingTail
    addSubview label
    label
  end
  
  def addButton label, x, y, w, size, action
    button = NSButton.alloc.initWithFrame([x, y, w, 20])
    button.setBezelStyle(NSRoundRectBezelStyle)
    button.setAutoresizingMask(NSViewMaxXMargin)
    button.setTitle(label)
    button.setTarget(self)
    button.setAction(action)
    fontSize = NSFont.systemFontSizeForControlSize(size)
    cell = button.cell
    font = NSFont.fontWithName_size(cell.font.fontName, fontSize)
    cell.setFont(font)
    cell.setControlSize(size)
    cell.setLineBreakMode(NSLineBreakByTruncatingTail)
    addSubview button
    button
  end
  
  def makeColumn(x, y, start)
    for i in 0..8 do
      @tracks[start + i] = addLabel x, y, 180, NSMiniControlSize, "songClicked:", "trackClicked:"
      y -= 13
    end
  end

  def setTrackImage image
    begin
      @delegate.track.artworks.objectAtIndex(0).data = image
      @delegate.curView.image = image
    rescue
    end
  end
  
  def setTrackAlbum str
    begin
      @delegate.track.album = str
      @delegate.curView.album = str
    rescue
    end
  end
  
  def setTrackArtist str
    begin
      @delegate.track.artist = str
      @delegate.curView.artist = str
    rescue
    end
  end
  
  def setTrackYear str
    begin
      @delegate.track.year = str
      @delegate.curView.year = str
    rescue
    end
  end
  
  def setTrackNumber disc, number
    begin
      @delegate.track.discNumber = disc
      @delegate.track.trackNumber = number
      @delegate.track.trackCount = nil
      @delegate.curView.trackNumber = number
    rescue
    end
  end
  
  def setTrackName name
    begin
      @delegate.track.name = name
      @delegate.curView.song = name
    rescue
    end
  end
  
  def allClicked(sender)
    imageClicked(sender)
    albumClicked(sender)
    artistClicked(sender)
    yearClicked(sender)
    songClicked(@tracks[@score_index]) if @score_index
  end

  def imageClicked(sender)
    setTrackImage @image.image if @image.image
  end
  
  def albumClicked(sender)
    setTrackAlbum @album.stringValue if @album.stringValue != ""
  end
  
  def artistClicked(sender)
    setTrackArtist @artist.stringValue if @artist.stringValue != ""
  end
  
  def yearClicked(sender)
    setTrackYear @year.stringValue if @year.stringValue != ""
  end
  
  def trackClicked(sender)
    str = sender.stringValue
    if str.to_s =~ /(([^-]+)-)?([0-9]+)\. (.*)/
      setTrackNumber discFrom($2), $3.to_i
    end
  end
  
  def songClicked(sender)
    str = sender.stringValue
    if str.to_s =~ /(([^-]+)-)?([0-9]+)\. (.*)/
      setTrackNumber discFrom($2), $3.to_i
      setTrackName cleanUpTrackName($4)
    end
  end
  
  def loadImage
    self.image = image_from(@imageURL)
  end

  def cleanUpTrackName(name)
    artist = @artist.stringValue
    if name =~ /(.*) - (.*)/
      if $2.include?(artist)
        $1
      else
        $&
      end
    else
      name
    end
  end
end

def discFrom str
  if str
    if str >= '0' and str <= '9'
      str.to_i
    else
      " abcdefg".index str.downcase
    end
  else
    nil
  end
end

File: trunk/AlbumView.m
//
//  AlbumView.m
//  ManiaDB
//
//  Created by Appledelhi on 11/27/08.
//  Copyright 2008 Appledelhi. All rights reserved.
//

#import "AlbumView.h"
#import "MyTextField.h"
#import "Util.h"
#import "AppDelegate.h"

@implementation AlbumView

- (NSTextField*)addLabel:(NSRect)frame size:(int)controlSize action:(SEL)action
{
  MyTextField* label = [[MyTextField alloc] initWithFrame:frame];
  [label setAutoresizingMask:NSViewWidthSizable];
  [label setStringValue:@""];
  [label setEditable:NO];
  [label setBezeled:NO];
  [label setDrawsBackground:NO];
  [label setTextColor:[NSColor whiteColor]];
  [label setTarget:self];
  [label setAction:action];
  float fontSize = [NSFont systemFontSizeForControlSize:controlSize];
  NSCell* cell = [label cell];
  NSFont* font = [NSFont fontWithName:[[cell font] fontName] size:fontSize];
  [cell setFont:font];
  [cell setControlSize:controlSize];
  [cell setLineBreakMode:NSLineBreakByTruncatingTail];
  [self addSubview:label];
  return label;
}

- (NSButton*)addButton:(NSString*)label frame:(NSRect)frame size:(int)controlSize
                action:(SEL)action
{
  NSButton* button = [[NSButton alloc] initWithFrame:frame];
  [button setBezelStyle:NSRoundRectBezelStyle];
  [button setAutoresizingMask:NSViewMaxXMargin];
  [button setTitle:label];
  [button setTarget:self];
  [button setAction:action];
  float fontSize = [NSFont systemFontSizeForControlSize:controlSize];
  NSCell* cell = [button cell];
  NSFont* font = [NSFont fontWithName:[[cell font] fontName] size:fontSize];
  [cell setFont:font];
  [cell setControlSize:controlSize];
  [cell setLineBreakMode:NSLineBreakByTruncatingTail];
  [self addSubview:button];
  return button;
}

- (void)makeColumnX:(float)x y:(float)y
{
  for (int i = 0; i < 9; ++i) {
    [_tracks addObject:[self addLabel:NSMakeRect(x, y, 180, 13) size:NSMiniControlSize
                               action:@selector(trackClicked:)]];
    y -= 13;
  }
}

- (id)initWithFrame:(NSRect)frame active:(BOOL)active delegate:(AppDelegate*)delegate
{
    self = [super initWithFrame:frame];
    if (self) {
      [self setAutoresizingMask:NSViewWidthSizable];
      _delegate = delegate;

      float buttonSize = frame.size.height - 10;
      if (active) {
        _image = [[NSButton alloc] initWithFrame:NSMakeRect(5, 10, buttonSize, buttonSize)];
        [_image setBezelStyle:NSRegularSquareBezelStyle];
        [_image setTitle:@""];
        [_image setTarget:self];
        [_image setAction:@selector(imageClicked:)];
      } else {
        _image = [[NSImageView alloc] initWithFrame:NSMakeRect(5, 10, buttonSize, buttonSize)];
        [(id)_image setImageFrameStyle:NSImageFrameGroove];
      }
      [[_image cell] setImageScaling:NSImageScaleProportionallyDown];
      [self addSubview:_image];
      
      _imageSize = [self addLabel:NSMakeRect(5, -10, buttonSize, 20) size:NSMiniControlSize action:nil];
      [_imageSize setAlignment:NSCenterTextAlignment];
      [_imageSize setAutoresizingMask:0];
      
      float x = 5 + buttonSize + 5;
      float y = 5 + buttonSize - 11;
      float w = frame.size.width - x - 5 - 45;

      _album = [self addLabel:NSMakeRect(x, y, w, 20) size:NSRegularControlSize
                       action:@selector(albumClicked:)];
      if (active)
        _all = [self addButton:@"Set" frame:NSMakeRect(x + 5, y + 2, 40, 20)
                          size:NSMiniControlSize action:@selector(allClicked:)];
      y -= 22;
      _artist = [self addLabel:NSMakeRect(x, y, 295, 20) size:NSSmallControlSize
                        action:@selector(artistClicked:)];
      _year = [self addLabel:NSMakeRect(x + 300, y, 50, 20) size:NSSmallControlSize
                      action:@selector(yearClicked:)];
      _tracks = [NSMutableArray array];
      y -= 10;
      for (int i = 0; i < 4; ++i) {
        [self makeColumnX:x + (180 + 5) * i y:y];
      }
      _index = 0;
      _score = 0;
      _scoreIndex = -1;
    }
    return self;
}

- (void)setAlbum:(NSString*)album
{
  [_album setStringValue:album];
  [_album sizeToFit];
  [_album setToolTip:album];
  NSRect frame = [_album frame];
  if (_all)
    [_all setFrameOrigin:NSMakePoint(frame.origin.x + frame.size.width + 10, frame.origin.y - 3)];
}

- (void)setImage:(NSImage*)image
{
  [_image setImage:image];
  if (image && [image size].width > 0)
    [_imageSize setStringValue:[NSString stringWithFormat:@"%dx%d", (int)[image size].width,
                                (int)[image size].height]];
  else
    [_imageSize setStringValue:@""];
  [_image setNeedsDisplay:YES];
}

- (void)setImageFrom:(NSString*)url
{
  [self setImage:imageFrom(url)];
}

- (void)setArtist:(NSString*)artist
{
  if (artist) {
    [_artist setStringValue:artist];
    [_artist setToolTip:artist];
  }
}

- (void)setYear:(int)year
{
  if (year > 0) {
    [_year setIntValue:year];
  }
}

- (void)setTrackNumber:(int)trackNumber
{
  _trackNumber = trackNumber;
  if (_trackNumber > 0) {
    [[_tracks objectAtIndex:0] setStringValue:_song ? format(@"%d. %@", _trackNumber, _song) :
                                                      format(@"%d.", _trackNumber)];
  } else if (_song) {
    [[_tracks objectAtIndex:0] setStringValue:_song];
  }
}

- (void)setSong:(NSString*)song
{
  _song = song;
  NSTextField* track = [_tracks objectAtIndex:0];
  NSString* str = _trackNumber > 0 ? format(@"%d. %@", _trackNumber, _song) : _song;
  [track setStringValue:str];
  [track setToolTip:str];
}

- (void)addSong:(NSString*)song disc:(NSString*)disc track:(NSString*)trackNo score:(int)score
{
  NSString* label = disc ? format(@"%@-%@. %@", disc, trackNo, song) : format(@"%@. %@", trackNo, song);
  if (_index < [_tracks count]) {
    NSTextField* track = [_tracks objectAtIndex:_index];
    [track setStringValue:label];
    [track setToolTip:label];
    if (score > _score) {
      if (_scoreIndex >= 0)
        [[_tracks objectAtIndex:_scoreIndex] setTextColor:[NSColor whiteColor]];
      [track setTextColor:[NSColor greenColor]];
      _trackNumber = [trackNo intValue];
      _song = song;
      _score = score;
      _scoreIndex = _index;
    }
    _index++;
  }
}

- (void)setTrackImage:(NSImage*)image
{
  [[[[_delegate track] artworks] objectAtIndex:0] setData:(id)image];
  [[_delegate curView] setImage:image];
}

- (void)setTrackAlbum:(NSString*)str
{
  [[_delegate track] setAlbum:str];
  [[_delegate curView] setAlbum:str];
}

- (void)setTrackArtist:(NSString*)str
{
  [[_delegate track] setArtist:str];
  [[_delegate curView] setArtist:str];
}

- (void)setTrackYear:(int)year
{
  [[_delegate track] setYear:year];
  [[_delegate curView] setYear:year];
}

- (void)setTrackDisc:(int)disc number:(int)num
{
  id track = [_delegate track];
  [track setDiscNumber:disc];
  [track setTrackNumber:num];
  [track setTrackCount:0];
  [[_delegate curView] setTrackNumber:num];
}

- (void)setTrackName:(NSString*)str
{
  [[_delegate track] setName:str];
  [[_delegate curView] setSong:str];
}

- (void)imageClicked:sender
{
  NSImage* image = [_image image];
  if (image) {
    [self setTrackImage:image];
  }
}

- (void)albumClicked:sender
{
  NSString* str = [_album stringValue];
  if ([str length] > 0) {
    [self setTrackAlbum:str];
  }
}

- (void)artistClicked:sender
{
  NSString* str = [_artist stringValue];
  if ([str length] > 0) {
    [self setTrackArtist:str];
  }
}

- (void)yearClicked:sender
{
  int year = [_year intValue];
  if (year > 0) {
    [self setTrackYear:year];
  }
}

static NSString* trackPattern = @"(([^-]+)-)?([0-9]+)\\. (.*)";

- (void)trackClicked:sender
{
  NSString* str = [sender stringValue];
  NSString* disc = match(str, trackPattern, 2);
  NSString* track = match(str, trackPattern, 3);
  NSString* name = match(str, trackPattern, 4);
  if (track && name) {
    [self setTrackDisc:discFrom(disc) number:[track intValue]];
    if ([sender nameClicked]) {
      [self setTrackName:cleanUpTrackName(name, [_artist stringValue])];
    }
  }
}

- (void)allClicked:sender
{
  [self imageClicked:sender];
  [self albumClicked:sender];
  [self artistClicked:sender];
  [self yearClicked:sender];
  if (_scoreIndex >= 0) {
    [self trackClicked:[_tracks objectAtIndex:_scoreIndex]];
  }
}

@end
